### ISSUE TITLE
Attribute handling with IDs, classes, and existing attributes buggy

### ISSUE THREAD

**Issue by phoerious** [lexbor/lexbor#139]
I noticed the following bugs when it comes to attributes:

1. Setting id or class attributes on a node that has been created by the HTML parser leads to a crash
2. Setting id or class attributes on an element created by `lxb_dom_document_create_element()` simply adds the attribute without checking if one already exists (resulting in duplicate attributes)
3. An id or class attribute added by `lxb_dom_document_create_element()` on such an element cannot be found by `lxb_dom_element_get_attribute()` or `lxb_dom_element_attr_is_exist()`
4. Serializing a tree where an existing "custom" attribute (other than id or class) has been replaced by `lxb_dom_element_set_attribute()` hangs forever (probably an infinite pointer loop)

Here is an example file demonstrating all the bugs:

```c
#include <lexbor/html/html.h>

void serialize(lxb_dom_element_t* node, lxb_dom_document_t* document) {
    lexbor_str_t* html_str = lexbor_str_create();
    lxb_html_serialize_tree_str(lxb_dom_interface_node(node), html_str);
    printf("%s\n", html_str->data);
    lexbor_str_destroy(html_str, document->text, true);
}

int main(void) {

    // 1. -------------------------------------------------------------------------

    static const lxb_char_t html[] = "<html><body id=\"foo\" class=\"bar\"></body></html>";
    lxb_html_document_t* html_document = lxb_html_document_create();
    lxb_html_document_parse(html_document, html, sizeof(html) - 1);
    lxb_dom_document_t* document = &html_document->dom_document;

    lxb_dom_element_t* body = lxb_dom_interface_element(lxb_html_document_body_element(html_document));

    // OK!
    // <body id="foo" class="bar"></body>
    serialize(body, document);

    const lxb_char_t* id_val = lxb_dom_element_get_attribute(body, "id", 2, NULL);
    const lxb_char_t* class_val = lxb_dom_element_get_attribute(body, "class", 5, NULL);

    // OK!
    // foo, bar
    printf("%s, %s\n", id_val, class_val);

    lxb_dom_attr_t* attr = lxb_dom_element_attr_by_name(body, "id", 2);
    //lxb_dom_attr_t* attr = lxb_dom_element_attr_is_exist(body, "id", 2);    // Also OK
    lxb_dom_attr_set_value(attr, "baz", 3);

    // OK!
    // <body id="baz" class="bar"></body>
    serialize(body, document);

    // CRASH!
    //lxb_dom_element_set_attribute(body, "id", 2, "baz", 3);
    //serialize(body, document);


    // 2. -------------------------------------------------------------------------

    lxb_dom_element_t* new_el = lxb_dom_document_create_element(document, "div", 3, NULL);
    lxb_dom_node_insert_child(lxb_dom_interface_node(body), lxb_dom_interface_node(new_el));

    // OK!
    // <body id="baz" class="bar"><div></div></body>
    serialize(body, document);

    lxb_dom_element_set_attribute(new_el, "id", 2, "baz", 3);

    // OK!
    // <body id="baz" class="bar"><div id="baz"></div></body>
    serialize(body, document);

    lxb_dom_element_set_attribute(new_el, "id", 2, "baz", 3);

    // WTF?!!!
    // <body id="baz" class="bar"><div id="baz" id="baz"></div></body>
    serialize(body, document);

    lxb_dom_element_set_attribute(new_el, "id", 2, "baz", 3);
    lxb_dom_element_set_attribute(new_el, "id", 2, "baz", 3);
    lxb_dom_element_set_attribute(new_el, "class", 5, "baz", 3);
    lxb_dom_element_set_attribute(new_el, "class", 5, "baz", 3);

    // ?????????!!!
    // <body id="baz" class="bar"><div id="baz" id="baz" id="baz" id="baz" class="baz" class="baz"></div></body>
    serialize(body, document);


    // 3. -------------------------------------------------------------------------

    // Cannot retrieve any of these ID attributes!
    if (lxb_dom_element_attr_is_exist(new_el, "id", 2) == NULL) {
        printf("No ID attribute (this is wrong)!\n");
    }


    // 4. -------------------------------------------------------------------------

    // Let's try that with other attributes
    lxb_dom_element_set_attribute(new_el, "lang", 4, "en", 2);

    // OK!
    // <body id="baz" class="bar"><div id="baz" id="baz" id="baz" id="baz" class="baz" class="baz" lang="en"></div></body>
    serialize(body, document);

    attr = lxb_dom_element_attr_is_exist(new_el, "lang", 4);
    lxb_dom_attr_set_value(attr, "de", 2);

    // OK!
    // <body id="baz" class="bar"><div id="baz" id="baz" id="baz" id="baz" class="baz" class="baz" lang="de"></div></body>
    serialize(body, document);

    lxb_dom_element_set_attribute(new_el, "lang", 4, "ru", 2);

    // HANGS FOREVER!
    serialize(body, document);

    return 0;
}
```

**Comment by phoerious** [comment]
Another opportunity where Lexbor can hang up indefinitely is when `lxb_dom_node_insert_before()` or `lxb_dom_node_insert_after()` a node that has a parent already. That should probably be fixed, too, although removing it first is an easy workaround, which I had to apply here: https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/develop/resiliparse/parse/html.pyx#L776

Unfortunately, for the issues with id and class attributes, there doesn't seem to be a suitable workaround.

**Comment by lexborisov** [comment]
@phoerious 

First example, my result:
```
<body id="foo" class="bar"></body>
foo, bar
<body id="baz" class="bar"></body>
<body id="baz" class="bar"><div></div></body>
<body id="baz" class="bar"><div id="baz"></div></body>
<body id="baz" class="bar"><div id="baz"></div></body>
<body id="baz" class="bar"><div id="baz" class="baz"></div></body>
<body id="baz" class="bar"><div id="baz" class="baz" lang="en"></div></body>
<body id="baz" class="bar"><div id="baz" class="baz" lang="de"></div></body>
<body id="baz" class="bar"><div id="baz" class="baz" lang="ru"></div></body>
```
no endless loops.

Are you using the latest version from github?

**Comment by phoerious** [comment]
Actually, no. Just noticed that I tested it on the latest stable release. I rebuilt Lexbor from Git and the issues seem to be fixed. Also commenting in lines 41 and 42 doesn't crash anymore.

I guess it's about time for a new release then. There are a lot of important fixes. My wrapper is also mostly finished and I am only waiting for the new Lexbor version now: https://resiliparse.chatnoir.eu/en/latest/man/parse/html.html
